### ISSUE TITLE
fix(cli): 'skipping empty file' mislabels tiny, binary, and UTF-16 skips

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#1287]
## Summary

`bca` prints `skipping empty file: <path>` for every file the library declines to read, but "empty" is only one of the reasons — 1–3-byte valid sources, multi-KB invalid-UTF-8/binary files, and UTF-16-BOM files all get the same message.

## Location

- `big-code-analysis-cli/src/dispatch.rs:185`
- `src/tools.rs:126-167` (`read_file_with_eol` doc: `Ok(None)` covers empty, too-small, undecodable bytes, and UTF-16 BE/LE BOM alike)

## Evidence

Found during the 2026-08-08 black-box audit sweep: a multi-kilobyte binary file and a 1-byte valid source (`x`) both produce `skipping empty file: …`. The library collapses all skip reasons into `Ok(None)`, so the CLI cannot name the real cause.

## Expected Behavior

The diagnostic names the actual reason (empty / too small / not valid UTF-8 / UTF-16 BOM), or at minimum uses wording that does not assert emptiness, e.g. `skipping unreadable or empty file`.

## Actual Behavior

Every `Ok(None)` is reported as an empty file. A user investigating why a real, non-empty source was skipped is told it is empty.

## Impact

Misleading diagnostics during triage of skipped files; adjacent to #1238 (the py bindings inherit the same unlabeled `Ok(None)` and silently drop batch slots).

## Note

The clean fix is library-side API expressiveness: `read_file_with_eol` returning a skip-reason enum (e.g. `Ok(Either<Vec<u8>, SkipReason>)`-shaped) that the CLI and py bindings can both render; #1238's fix plan touches the same seam, so coordinate.

---

## Resolution Plan

`read_file_with_eol` is public, re-exported API, so its `io::Result<Option<Vec<u8>>>` shape cannot change within 2.x (STABILITY.md). Fix via an additive sibling that names the reason, and keep the old function as a shim.

1. **Library** (`src/tools.rs` + `src/lib.rs` re-export): add a public `SkipReason` enum — `Empty` (size 0), `TooSmall` (1-3 bytes, and the shrank-during-probe short-read case), `NotUtf8` (failed probe), `Utf16Bom` (BE/LE BOM) — and a new public function (e.g. `read_file_with_eol_classified`) returning `io::Result<Result<Vec<u8>, SkipReason>>`. Each existing `return Ok(None)` branch already knows its reason; the classification is free. Reimplement `read_file_with_eol` as a one-line delegation (`.map(|r| r.ok())`-shaped) so behaviour is provably unchanged; do not deprecate it in this change.
2. **CLI** (`big-code-analysis-cli/src/dispatch.rs:185`): switch to the classified call and render a per-reason message via `warn` (lowercase, bare message per the diagnostic-prefix convention): keep `skipping empty file: <p>` for `Empty`; add `skipping tiny file (3 bytes or fewer): <p>`, `skipping non-UTF-8 file: <p>`, `skipping UTF-16 file (re-encode as UTF-8): <p>` — exact wording bikesheddable in review, but no message may assert emptiness for a non-empty file.
3. **Coordinate the other two front-ends on the same seam**:
   - py (`big-code-analysis-py/src/analysis.rs`): #1238's fix touches this exact call; switching it to the classified variant now (still returning `Ok(None)` upward) costs one line and gives #1238's placeholder/None slots a reason to attach later. Do not widen this issue into #1238's contract change — just leave the seam ready.
   - web (`big-code-analysis-web/src/web/server/handlers.rs`): check how it reports the `None` case and align wording only if it also claims "empty".
4. **Tests**:
   - `src/tools_tests.rs`: unit tests pinning each `SkipReason` variant from a crafted file (0-byte, 1-byte, binary head, UTF-16 BOM) plus one asserting the shim's `Option` equivalence.
   - `big-code-analysis-cli/tests/discovery/warning_flag.rs:65`: update the pinned `warning: skipping empty file:` expectation and add cases for a 1-byte file and a binary file asserting the new reason-specific strings (test-via-revert: they fail with the old single message).
5. **Docs/changelog**: update the `read_file_with_eol` doc paragraph to point at the classified variant; CHANGELOG `## [Unreleased]` additive-API + CLI-diagnostic entry; cross-check STABILITY.md's public-surface list mentions the new export. New enum should be `#[non_exhaustive]` so a future reason (e.g. a BOM family addition) lands additively.
6. **Validation**: `make pre-commit`; manual spot-check of the audit sweep's two reproducers (multi-KB binary, 1-byte `x`).

## Assessment

| Dimension  | Rating |
|------------|--------|
| Difficulty | Medium |
| Complexity | Medium |
| Priority   | Medium |

## Resolution

Fixed on `fix/issue-1287` (commit `77a6743e`). Not closed — the
integration branch owns that.

The library gained an additive classified reader; the CLI renders its
reason.

- `SkipReason` (`#[non_exhaustive]`, `Display`) with one variant per
  real skip branch: `Empty`, `TooSmall` (1–3 bytes *and* the
  shrank-during-probe short read), `Utf16Bom`, `NotUtf8`.
- `read_file_with_eol_classified(path) -> io::Result<Result<Vec<u8>,
  SkipReason>>`.
- `read_file_with_eol` is unchanged in shape and is now literally
  `read_file_with_eol_classified(path).map(Result::ok)`, so shim
  equivalence holds by construction.
- The CLI's skip warning is now `skipping {reason}: {path}`; only an
  actually-empty file reads "empty file". The severity prefix stays in
  `warn`.
- Internally the read body is a private `read_gated` whose two failure
  modes ride one `ReadStop` enum, so each gate is a `?` rather than a
  hand-wrapped nested-`Result` early return. It measures below the
  pre-change function (halstead.effort 37061 vs 41378); the baseline
  `nexits` entry is renamed, not raised.

The web crate reads no files, so nothing there claimed emptiness. The py
bindings are deliberately untouched: surfacing the reason through
`analyze`/`batch` is a Python-contract change on top of #1238's held
slot, not a one-line swap. Worth a follow-up issue if the batch API
should say *why* a slot is `None`.

Docs: `docs/file-detection.md` (reader table + `SkipReason` paragraph)
and `STABILITY.md` (new **File readers** bullet; the per-variant
`Display` wording is explicitly not SemVer-protected).

Tests: six-fixture branch table plus shim-equivalence and wording tests
in `src/tools_tests.rs`; three reason-specific CLI cases in
`tests/discovery/warning_flag.rs`. Every one verified by test-via-revert
— see the comment thread for the perturbation table.

**Comment by dekobon** [comment]
## General Plan for Fix

### Root Cause

`read_file_with_eol` (`src/tools.rs`) collapses four distinct skip reasons (empty, ≤3 bytes, undecodable bytes, UTF-16 BOM) into one unlabeled `Ok(None)`, so callers cannot render an accurate diagnostic; the CLI hardcodes the "empty" wording.

### Implementation Steps

1. Introduce a `SkipReason` enum in `src/tools.rs` (`Empty`, `TooSmall`, `Undecodable`, `Utf16Bom`) and change `read_file_with_eol` to return `Ok(Result<Vec<u8>, SkipReason>)`-shaped data (or a dedicated `ReadOutcome` enum) — keep the existing `Ok(None)`-equivalent path semantics unchanged.
2. Update the CLI call site (`big-code-analysis-cli/src/dispatch.rs:185`) to render the reason: `skipping <reason> file: <path>`.
3. Update the py bindings' batch path (#1238 touches the same seam) to carry the reason into its per-file result slot.
4. Check STABILITY.md: `read_file_with_eol` visibility — if it is `pub`, the shape change is breaking and must be recorded **(breaking)** in CHANGELOG under Unreleased, or an additive sibling fn added instead.

### Tests to Add/Update

- Unit tests in `src/tools.rs` pinning each `SkipReason` for: empty file, 1-byte file, invalid-UTF-8 bytes, UTF-16 LE/BE BOM.
- CLI test asserting the rendered message names the right reason for a binary file (not "empty").

### Verification

- [ ] `cargo test -p big-code-analysis -p big-code-analysis-cli` passes
- [ ] `cargo clippy --workspace --all-targets -- -D warnings` clean
- [ ] Manual: `bca metrics <binary-file>` prints a reason other than "empty"

**Comment by dekobon** [comment]
Added resolution plan and assessment ratings (Difficulty: Medium, Complexity: Medium, Priority: Medium). Key constraint found during investigation: read_file_with_eol is public re-exported API, so the skip-reason enum must arrive via an additive sibling function (old function becomes a shim) rather than a return-type change — which also hands #1238's py-binding fix the seam it needs. No labels applied.

**Comment by dekobon** [comment]
## Fixed on `fix/issue-1287` (commit 77a6743e)

Reproduced first, against the tree as it stands on the integration
branch (the issue's line numbers had drifted — `dispatch.rs:185` is now
`validate_and_resolve_file`, reshaped by the #1303 fix):

```
warning: skipping empty file: /tmp/…/tiny.rs      # 1 byte, `x`
warning: skipping empty file: /tmp/…/binary.rs    # 4 KB of /dev/urandom
warning: skipping empty file: /tmp/…/empty.rs     # 0 bytes
warning: skipping empty file: /tmp/…/u16.rs       # UTF-16 LE BOM
```

After:

```
warning: skipping empty file: /tmp/…/empty.rs
warning: skipping file too small to analyze (3 bytes or fewer): /tmp/…/tiny.rs
warning: skipping file with non-UTF-8 contents: /tmp/…/binary.rs
warning: skipping UTF-16 file (unsupported encoding): /tmp/…/u16.rs
```

### Library (additive, per the plan)

`read_file_with_eol`'s `io::Result<Option<Vec<u8>>>` is public 2.x API
and keeps its shape. Added alongside it:

- `SkipReason` — `#[non_exhaustive]`, `Copy`, `Display`. Four variants,
  mapped 1:1 against the real `Ok(None)` branches in the code rather
  than the plan's list: `Empty` (size 0), `TooSmall` (1–3 bytes **and**
  the shrank-during-probe `UnexpectedEof` short read), `Utf16Bom`,
  `NotUtf8`. Named `NotUtf8` rather than the comment thread's
  `Undecodable`: `Utf16Bom` is equally undecodable, so the more
  specific name is the one that discriminates.
- `read_file_with_eol_classified` — `io::Result<Result<Vec<u8>,
  SkipReason>>`. Outer `Result` is the I/O outcome, inner is the
  analysis outcome.
- `read_file_with_eol` is now literally
  `read_file_with_eol_classified(path).map(Result::ok)`, so shim
  equivalence is by construction and not by parallel maintenance.

`SkipReason: Display` renders a noun phrase (`empty file`, `file too
small to analyze (3 bytes or fewer)`, …) with no severity word, so the
CLI's `skipping {reason}: {path}` keeps the prefix in `warn` per the
diagnostic convention.

The read body moved to a private `read_gated` returning `Result<Vec<u8>,
ReadStop>`, where `ReadStop` is `Io(io::Error) | Skip(SkipReason)` with
`From` impls. That lets each gate be spelled `?` instead of
hand-wrapping four early returns in the public nested-`Result` shape,
and it measures *below* the pre-change function (halstead.effort 37061
vs 41378) — the classification costs no complexity. The
`.bca-baseline.toml` `nexits = 7` entry is renamed to `read_gated`, same
value.

### Other front-ends

- **web**: never reads files off disk (all payloads are in-memory); the
  only `read_file_with_eol` mention there is a comment about matching
  its EOL normalisation. Nothing to align.
- **py**: left untouched deliberately. `analyze_path` returns `Ok(None)`
  and `batch` turns that into a held `None` slot (#1238, landed as
  010b3fa1 + f6c144e1). Surfacing the reason there is a Python-contract
  change — a new field or error class — not a one-line swap, so the seam
  is left ready rather than widened. Worth its own issue if the batch
  API should report *why* a slot is empty.

### Tests

Library (`src/tools_tests.rs`): a six-fixture table pinning every branch
(0 bytes, 1 byte, 3 bytes, UTF-16 LE, UTF-16 BE, stray `0xFF` padded past
the 64-byte probe), each asserting both the classified reason and that
the shim still reports a skip; an accepted-file case asserting both entry
points agree on CRLF-normalised bytes (so a shim returning `None`
unconditionally could not pass); and a wording test that only `Empty` may
claim emptiness.

CLI (`tests/discovery/warning_flag.rs`): three new cases asserting the
reason-specific string **and** `contains("empty").not()`.

Test-via-revert, all with in-memory backups:

| perturbation | result |
| --- | --- |
| CLI message back to the single `skipping empty file` | the 3 new CLI tests fail, `warning_flag_emits_empty_file` still passes |
| `for_size` swaps `Empty`/`TooSmall` | branch-table test fails, only it |
| `Utf16Bom` → `NotUtf8` in the probe | branch-table test fails, only it |
| `TooSmall`'s phrase → `"empty file"` | wording test fails, only it |

### Docs

`docs/file-detection.md` gains the classified reader row and a paragraph
on `SkipReason`; `STABILITY.md` gains a **File readers** bullet recording
the additive pair, the `#[non_exhaustive]` contract, and that the
per-variant `Display` wording is explicitly *not* SemVer-protected.

`make pre-commit`: `BCA_GATE: pass (gate=pre-commit)`.
